Libraries/Opc.Ua.Security.Certificates/X509Crl/X509Crl.cs

```diff
@@ -0,0 +1,345 @@
+            }
+        }
+
+        /// <inheritdoc/>
+        public IList<RevokedCertificate> RevokedCertificates
+        {
+            get
+            {
+                EnsureDecoded();
+                return m_revokedCertificates.AsReadOnly();
+            }
+        }
+
+        /// <inheritdoc/>
+        public X509ExtensionCollection CrlExtensions
+        {
+            get
+            {
+                EnsureDecoded();
+                return m_crlExtensions;
+            }
+        }
+
+        /// <inheritdoc/>
+        public byte[] RawData { get; private set; }
+        #endregion
+
+        #region Public Methods
+        /// <summary>
+        /// Verifies the signature on the CRL.
+        /// </summary>
+        public bool VerifySignature(X509Certificate2 issuer, bool throwOnError)
+        {
+            bool result;
+            try
+            {
+                var signature = new X509Signature(RawData);
+                result = signature.Verify(issuer);
+            }
+            catch (Exception)
+            {
+                result = false;
+            }
+            if (!result && throwOnError)
+            {
+                throw new CryptographicException("Could not verify signature on CRL.");
+            }
+            return result;
+        }
+
+        /// <summary>
+        /// Returns true if the certificate is revoked in the CRL.
+        /// </summary>
+        public bool IsRevoked(X509Certificate2 certificate)
+        {
+            if (certificate.IssuerName.Equals(IssuerName))
+            {
+                throw new CryptographicException("Certificate was not created by the CRL Issuer.");
+            }
+            EnsureDecoded();
+            var serialnumber = certificate.GetSerialNumber();
+            foreach (var revokedCert in RevokedCertificates)
+            {
+                if (serialnumber.SequenceEqual<byte>(revokedCert.UserCertificate))
+                {
+                    return true;
+                }
+            }
+            return false;
+        }
+        #endregion
+
+        #region Private Methods
+        /// <summary>
+        /// Decode the complete CRL.
+        /// </summary>
+        /// <param name="crl">The raw signed CRL</param>
+        internal void Decode(byte[] crl)
+        {
+            // Decode the Tbs and signature
+            m_signature = new X509Signature(crl);
+            // Decode the TbsCertList
+            DecodeCrl(m_signature.Tbs);
+        }
+
+        /// <summary>
+        /// Decode the Tbs of the CRL.
+        /// </summary>
+        /// <param name="tbs">The raw TbsCertList of the CRL.</param>
+        internal void DecodeCrl(byte[] tbs)
+        {
+            try
+            {
+                AsnReader crlReader = new AsnReader(tbs, AsnEncodingRules.DER);
+                var tag = Asn1Tag.Sequence;
+                var seqReader = crlReader?.ReadSequence(tag);
+                if (seqReader != null)
```

Comment:
Every time you think you've finished an item (such as after `crlReader.ReadSequence`, since the CRL should only be one sequence) you should call ThrowIfNotEmpty to be defensive against ill-formed data; otherwise you'll run into compat risks if you move off of your own reader (since a BCL version will definitely throw on extra data).

Reply:
great to know, I added throwifnotempty to all AsnReader and ReadSequence